searx/network/__init__.py

```diff
@@ -1,266 +1,483 @@
-    future = asyncio.run_coroutine_threadsafe(stream_chunk_to_queue(network, queue, method, url, **kwargs), get_loop())
 
-    # yield chunks
-    obj_or_exception = queue.get()
-    while obj_or_exception is not None:
-        if isinstance(obj_or_exception, Exception):
-            raise obj_or_exception
-        yield obj_or_exception
-        obj_or_exception = queue.get()
-    future.result()
+def provide_networkcontext(
+    network_name: Optional[str] = None, timeout: Optional[float] = None, start_time: Optional[float] = None
+):
+    """Set the NetworkContext, then call the wrapped function using searx.network.context.NetworkContext.call
 
+    The timeout is for the whole function and is infinite by default (None).
+    The timeout is counted from the current time or start_time if different from None
 
-def _close_response_method(self):
-    asyncio.run_coroutine_threadsafe(self.aclose(), get_loop())
-    # reach the end of _self.generator ( _stream_generator ) to an avoid memory leak.
-    # it makes sure that :
-    # * the httpx response is closed (see the stream_chunk_to_queue function)
-    # * to call future.result() in _stream_generator
-    for _ in self._generator:  # pylint: disable=protected-access
-        continue
+    Intended usage: to provide a NetworkContext for scripts in searxng_extra.
 
+    Example of usage:
 
-def stream(method, url, **kwargs) -> Tuple[httpx.Response, Iterable[bytes]]:
-    """Replace httpx.stream.
+    ```python
+    from time import sleep
+    from searx import network
 
-    Usage:
-    response, stream = poolrequests.stream(...)
-    for chunk in stream:
-        ...
+    @network.provide_networkcontext(timeout=3.0)
+    def main()
+        sleep(1.0)
+        # the timeout is automatically set to 2.0 (the remaining time for the NetworkContext).
+        my_ip = network.get("https://ifconfig.me/ip").text
+        print(my_ip)
 
-    httpx.Client.stream requires to write the httpx.HTTPTransport version of the
-    the httpx.AsyncHTTPTransport declared above.
+    if __name__ == '__main__':
+        main()
+    ```
     """
-    generator = _stream_generator(method, url, **kwargs)
-
-    # yield response
-    response = next(generator)  # pylint: disable=stop-iteration-return
-    if isinstance(response, Exception):
-        raise response
 
-    response._generator = generator  # pylint: disable=protected-access
-    response.close = MethodType(_close_response_method, response)
-
-    return response, generator
+    def func_outer(func: Callable[P, R]):
+        @wraps(func)
+        def func_inner(*args: P.args, **kwargs: P.kwargs) -> R:
+            with networkcontext_for_thread(network_name, timeout, start_time) as network_context:
+                return network_context.call(func, *args, **kwargs)
+
+        return func_inner
+
+    return func_outer
+
+
+def request(
+    method: str,
+    url: str,
+    params: Optional[httpx._types.QueryParamTypes] = None,
```

Comment:
FYI (no need to change) with `from __future__ import annotations` you can also write:

```python
    params: httpx._types.QueryParamTypes | None = None
```